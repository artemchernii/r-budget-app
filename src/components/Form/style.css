/* Form input */
/* Style the container to create a floating label effect */
form {
    padding: 20px;
    margin-left: 40px;
    width: 60%;
    border: 3px solid #fffffe;
    border-radius: 10px;
}
.input-container {
    position: relative;
    margin-bottom: 20px;
}

/* Style the input field */
.material-input {
    font-size: 18px;
    padding: 10px;
    display: block;
    width: 100%;
    border: none;
    border-bottom: 1px solid #ccc;
    background-color: transparent;
    transition: border-bottom-color 0.3s;
}

.material-input:focus {
    outline: none;
    border-bottom-color: #ccc; /* Highlight color when focused */
}

/* Style the label */
.material-label {
    position: absolute;
    left: 10px;
    top: 10px;
    font-size: 16px;
    color: #888;
    pointer-events: none;
    transition: top 0.3s, font-size 0.3s;
}

/* Shrink the label when input is focused or has value */
.material-input:focus + .material-label,
.material-input:not(:placeholder-shown) + .material-label {
    top: -10px;
    font-size: 14px;
    color: #ccc; /* Highlight color when focused */
}
