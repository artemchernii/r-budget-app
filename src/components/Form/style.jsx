import styled from 'styled-components';
import { THEMES } from '../../providers/themes/themeList';

const FormWrapper = styled.form`
    padding: 20px;
    width: 50%;
    border: 1px solid ${({ theme }) => theme.color};
    border-radius: 3px;

    .input-container {
        position: relative;
        margin-bottom: 20px;
    }

    /* Style the input field */
    .material-input {
        font-size: 18px;
        padding: 10px;
        display: block;
        width: 100%;
        border: none;
        border-bottom: 1px solid #ccc;
        background-color: transparent;
        transition: border-bottom-color 0.3s;
        color: ${({ theme }) => theme.color};
    }

    .material-input:focus {
        outline: none;
        border-bottom-color: #ccc;
    }

    .material-label {
        position: absolute;
        left: 10px;
        top: 10px;
        font-size: 16px;
        color: ${({ currentTheme }) =>
            currentTheme === THEMES.LIGHT ? '#000' : '#888'};
        pointer-events: none;
        transition: top 0.3s, font-size 0.3s;
    }

    .material-input:focus + .material-label,
    .material-input:not(:placeholder-shown) + .material-label {
        top: -10px;
        font-size: 14px;
        color: ${({ currentTheme }) =>
            currentTheme === THEMES.LIGHT ? '#000' : '#888'};
    }
`;

FormWrapper.displayName = 'FormWrapper';

export { FormWrapper };
