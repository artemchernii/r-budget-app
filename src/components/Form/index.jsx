import propTypes from 'prop-types';
import { useContext, useState } from 'react';
import { FormWrapper } from './style';
import { stateContext } from '../../providers/context/defaultContext';
import { Button } from '../App/style';

export default function Form({ handleSubmit }) {
    const [form, setForm] = useState({
        balance: 0,
        date: new Date().toISOString().substring(0, 10),
        comment: '',
    });

    const {
        state: { theme },
    } = useContext(stateContext);

    const onSubmit = (e) => {
        e.preventDefault();
        handleSubmit(form);
        setTimeout(() => {
            setForm({
                balance: 0,
                date: new Date().toISOString().substring(0, 10),
                comment: '',
            });
        }, 400);
    };
    const handleBalanceChange = (e) => {
        const { name, value } = e.target;
        setForm((c) => {
            return {
                ...c,
                [name]: value,
                // [name]: type === 'number' ? +value : value,
            };
        });
    };

    return (
        <FormWrapper currentTheme={theme} onSubmit={onSubmit} title="test-form">
            <div className="input-container">
                <input
                    type="date"
                    className="material-input"
                    required
                    name="date"
                    value={form.date}
                    onChange={handleBalanceChange}
                />
                <label className="material-label">Date</label>
            </div>
            <div className="input-container">
                <input
                    type="number"
                    className="material-input"
                    required
                    name="balance"
                    value={form.balance}
                    onChange={handleBalanceChange}
                    id="balance"
                />
                <label htmlFor="balance" className="material-label">
                    New sum
                </label>
            </div>
            <div className="input-container">
                <textarea
                    className="material-input"
                    required
                    name="comment"
                    value={form.comment}
                    onChange={handleBalanceChange}
                />
                <label className="material-label">Comment</label>
            </div>
            <Button type="submit">Save</Button>
        </FormWrapper>
    );
}

Form.propTypes = {
    handleSubmit: propTypes.func,
};
Form.defaultProp = {
    handleSubmit: () => {},
};
