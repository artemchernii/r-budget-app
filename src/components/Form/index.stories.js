import Form from './index';
import '../../index.css';

export default {
  title: 'Example/Form',
  component: Form,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
};

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default = {
  args: {
    onChange: () => {}
  },
};
