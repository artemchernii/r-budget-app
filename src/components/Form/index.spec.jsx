import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import Form from './index';

describe('Form component', () => {
    test('should show transaction', () => {
        const props = {
            value: 23,
            comment: 'Test comment',
            date: '01.01.2022',
        };
        const form = render(<Form {...props} />);
        expect(form).toMatchSnapshot();
    });
});
