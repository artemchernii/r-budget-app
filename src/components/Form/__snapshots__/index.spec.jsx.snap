// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form component > should show transaction 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="input-container"
        >
          <input
            class="material-input"
            name="date"
            required=""
            type="date"
            value="2023-11-06"
          />
          <label
            class="material-label"
          >
            Date
          </label>
        </div>
        <div
          class="input-container"
        >
          <input
            class="material-input"
            name="balance"
            required=""
            type="number"
            value="0"
          />
          <label
            class="material-label"
          >
            New sum
          </label>
        </div>
        <div
          class="input-container"
        >
          <textarea
            class="material-input"
            name="comment"
            required=""
          />
          <label
            class="material-label"
          >
            Comment
          </label>
        </div>
        <button
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="input-container"
      >
        <input
          class="material-input"
          name="date"
          required=""
          type="date"
          value="2023-11-06"
        />
        <label
          class="material-label"
        >
          Date
        </label>
      </div>
      <div
        class="input-container"
      >
        <input
          class="material-input"
          name="balance"
          required=""
          type="number"
          value="0"
        />
        <label
          class="material-label"
        >
          New sum
        </label>
      </div>
      <div
        class="input-container"
      >
        <textarea
          class="material-input"
          name="comment"
          required=""
        />
        <label
          class="material-label"
        >
          Comment
        </label>
      </div>
      <button
        type="submit"
      >
        Save
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
